crates/c-api/include/wasmtime/component.h

```diff
@@ -0,0 +1,635 @@
+ * #wasmtime_component_type_t
+ *
+ * \param val the #wasmtime_component_type_t to delete
+ */
+void wasmtime_component_type_delete(wasmtime_component_type_t *ty);
+
+/// Representation of a component in the component model.
+typedef struct wasmtime_component_t wasmtime_component_t;
+
+/**
+ * \brief Compiles a WebAssembly binary into a #wasmtime_component_t
+ *
+ * This function will compile a WebAssembly binary into an owned
+ #wasmtime_component_t.
+ *
+ * It requires a component binary, such as what is produced by Rust `cargo
+ component` tooling.
+ *
+ * This function does not take ownership of any of its arguments, but the
+ * returned error and component are owned by the caller.
+
+ * \param engine the #wasm_engine_t that will create the component
+ * \param buf the address of the buffer containing the WebAssembly binary
+ * \param len the length of the buffer containing the WebAssembly binary
+ * \param component_out on success, contains the address of the created
+ *        component
+ *
+ * \return NULL on success, else a #wasmtime_error_t describing the error
+ */
+wasmtime_error_t *
+wasmtime_component_from_binary(const wasm_engine_t *engine, const uint8_t *buf,
+                               size_t len,
+                               wasmtime_component_t **component_out);
+
+/**
+ * \brief Deletes a #wasmtime_component_t created by
+ * #wasmtime_component_from_binary
+ *
+ * \param component the component to delete
+ */
+void wasmtime_component_delete(wasmtime_component_t *component);
+
+/**
+ * \brief Representation of a component linker
+ *
+ * This type corresponds to a `wasmtime::component::Linker`.
+ *
+ * Due to the interaction between `wasmtime::component::Linker` and
+ * `wasmtime::component::LinkerInstance`, the latter being more of a builder,
+ * it is expected to first define the host functions through calls to
+ * #wasmtime_component_linker_define_func, then call
+ * #wasmtime_component_linker_build to create and populate the root
+ * `wasmtime::component::LinkerInstance` and it's descendants (if any).
+ */
+typedef struct wasmtime_component_linker_t wasmtime_component_linker_t;
+
+/**
+ * \brief Creates a new #wasmtime_component_linker_t for the specified engine.
+ *
+ * \param engine the compilation environment and configuration
+ *
+ * \return a pointer to the newly created #wasmtime_component_linker_t
+ */
+wasmtime_component_linker_t *
+wasmtime_component_linker_new(const wasm_engine_t *engine);
+
+/**
+ * \brief Deletes a #wasmtime_component_linker_t created by
+ * #wasmtime_component_linker_new
+ *
+ * \param linker the #wasmtime_component_linker_t to delete
+ */
+void wasmtime_component_linker_delete(wasmtime_component_linker_t *linker);
+
+/// Representation of a component instance
+typedef struct wasmtime_component_instance_t wasmtime_component_instance_t;
+
+// declaration from store.h
+typedef struct wasmtime_context wasmtime_context_t;
```

Comment:
I tend to favor forward declarations when possible, mostly for build time reasons (more to include is slower, and changes in the included header imply a re-build that could be avoided). What are the reasons for preferring a full `#include` ?

Reply:
AFAIK build-time concerns matter a lot for C++ but far less so for C, so I'd prefer to not duplicate definitions because structures like this can change over time.

Reply:
Plus one - C is super cheap to build and this is a trivial amount of just declarations.